frontend/simple/views/components/Chatroom/ChatMain.vue

```diff
@@ -0,0 +1,390 @@
+
+export default {
+  name: 'ChatMain',
+  components: {
+    MainHeader,
+    Avatar,
+    Loading,
+    List,
+    MenuParent,
+    MenuTrigger,
+    MenuContent,
+    MenuHeader,
+    MenuItem,
+    Message,
+    MessageInteractive,
+    MessageNotification,
+    ConversationGreetings,
+    SendArea
+  },
+  props: {
+    summary: {
+      type: Object, // { type: String, title: String, description: String, routerBack: String }
+      default: {}
+    },
+    details: {
+      type: Object, // { isLoading: Bool, conversation: Array, participants: Object }
+      default: {}
+    }
+  },
+  data () {
+    return {
+      config: {
+        isPhone: null
+      },
+      ephemeral: {
+        bodyPaddingBottom: '',
+        conversationIsLoading: false,
+        pendingMessages: []
+      }
+    }
+  },
+  created () {
+    // TODO #492 create a global Vue Responsive just for media queries.
+    var mediaIsPhone = window.matchMedia('screen and (max-width: 639px)')
+    this.config.isPhone = mediaIsPhone.matches
+    mediaIsPhone.onchange = (e) => { this.config.isPhone = e.matches }
+  },
+  updated () {
+    if (this.summary.title) {
+      // force conversation viewport to be at the bottom (most recent messages)
+      if (this.config.isPhone) {
+        setTimeout(function () {
+          window.scroll(0, document.body.offsetHeight)
+        }, 0) // REVIEW: without timeout it scrolls before the DOM is ready... why?
+      } else {
+        this.$refs.conversation && this.$refs.conversation.scroll(0, this.$refs.conversation.scrollHeight)
+      }
+    }
+  },
+  computed: {
+    bodyStyles () {
+      return this.config.isPhone ? { paddingBottom: this.ephemeral.bodyPaddingBottom } : {}
+    },
+    startedUnreadIndex () {
+      return this.details.conversation.findIndex(message => message.unread === true)
+    },
+    getMessageAt () {
+      let isCurrentUser
+
+      return this.details.conversation.map((message, index) => {
+        isCurrentUser = this.currentUserAttr.id === message.from
+
+        if (message.from === messageTypes.INTERACTIVE) {
+          return {
+            is: MessageInteractive,
+            id: message.id
+          }
+        }
+        if (message.from === messageTypes.NOTIFICATION) {
+          return {
+            is: MessageNotification,
+            text: message.text
+          }
+        }
+
+        return {
+          is: Message,
+          text: message.text,
+          who: this.who(isCurrentUser, message.from),
+          avatar: this.avatar(isCurrentUser, message.from),
+          variant: this.variant(isCurrentUser),
+          hideWho: this.shouldHideWho(index),
+          isSameSender: this.isSameSender(index)
+        }
+      })
+    },
+    getPendingAt () {
+      return this.ephemeral.pendingMessages.map((message, index) => ({
+        text: message.text,
+        who: this.who(true),
+        avatar: this.avatar(true),
+        variant: message.hasFailed ? 'failed' : 'sent',
```

Comment:
Inline string constants that need to be converted to constants.

Since these refer to something that is used in (and by) `Message.vue`, I recommend defining and exporting those constant in `Messages.vue` and importing them in this file to be reference here and anywhere else.

Reply:
Done, created `messageVariantTypes.js` with an object of constants.

A question: Can't we do named exports in a vue file? I tried to export those types at `Messages.vue` but it didn't work...